crates/oxc_linter/src/lint_adapter.rs

```diff
@@ -0,0 +1,1208 @@
+            Vertex::$alternate_vertex(expr) => matches!(expr, $alternate_vertex::$typed(..)),
+            _ => false,
+        }
+    };
+}
+
+macro_rules! coerce_ast_or_x {
+    ($contexts:ident, $alternate_vertex:ident, $typed:ident) => {
+        resolve_coercion_with($contexts, |v| match v {
+            Vertex::AstNode(node) => matches!(node.kind(), AstKind::$typed(..)),
+            Vertex::$alternate_vertex(expr) => matches!(expr, $alternate_vertex::$typed(..)),
+            _ => false,
+        })
+    };
+}
+
+macro_rules! get_ast_or_x_or_y {
+    ($v:ident, $x:ident, $y:ident, $typed:ident) => {
+        match $v {
+            Vertex::AstNode(node) => match node.kind() {
+                AstKind::$typed(be) => be,
+                _ => unreachable!(),
+            },
+            Vertex::$x(expr, ..) => match expr {
+                $x::$typed(be) => &be.0,
+                _ => unreachable!(),
+            },
+            Vertex::$y(expr2, ..) => match expr2 {
+                $y::$typed(be) => &be.0,
+                _ => unreachable!(),
+            },
+            _ => unreachable!(),
+        }
+    };
+}
+
+macro_rules! coerce_ast_or_x_or_y {
+    ($contexts:ident, $alternate_x:ident, $alternate_y:ident, $typed:ident) => {
+        resolve_coercion_with($contexts, |v| match v {
+            Vertex::AstNode(node) => matches!(node.kind(), AstKind::$typed(..)),
+            Vertex::$alternate_x(expr, ..) => matches!(expr, $alternate_x::$typed(..)),
+            Vertex::$alternate_y(expr, ..) => matches!(expr, $alternate_y::$typed(..)),
+            _ => unreachable!(),
+        })
+    };
+}
+
+macro_rules! string_matcher {
+    ($Self:ident, $t:ident, $expr:expr) => {
+        match $expr {
+            $t::StringLiteral(slit) => slit.value.as_str().into(),
+            $t::TemplateLiteral(tlit) => $Self::field_value_from_template_lit(tlit),
+            _ => FieldValue::Null,
+        }
+    };
+}
+
+macro_rules! ast_node_or_expression_field {
+    ($self:ident, $typed:ident, $attr:ident) => {
+        match $self {
+            Vertex::AstNode(node) => match node.kind() {
+                AstKind::$typed(be) => &be.$attr,
+                _ => unreachable!(),
+            },
+            Vertex::Expression(expr) => match expr {
+                Expression::$typed(be) => &be.$attr,
+                _ => unreachable!(),
+            },
+            _ => unreachable!(),
+        }
+    };
+}
+
+macro_rules! once_vertex_iter_if_some {
+    ($maybe_some:expr, $vertex_type:ident) => {
+        if let Some(some) = &$maybe_some {
+            Box::new(std::iter::once(Vertex::$vertex_type(&some.0)))
+        } else {
+            Box::new(std::iter::empty())
+        }
+    };
+}
+
+impl<'a> LintAdapter<'a> {
+    fn get_ast_node_iter(&self) -> VertexIterator<'a, Vertex<'a>> {
+        Box::new(self.semantic.nodes().nodes.clone().into_iter().map(|v| Vertex::AstNode(v)))
```

Comment:
At some point, eliminating this `.clone()` will probably improve performance further. Not a blocker for right now.